File: python/matrix.py
import sys
import os
from time import sleep
import RPi.GPIO as GPIO

LINES = 16

# 1 - OFF, 0 - ON
matrix = [
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0]
]

class Matrix:

    def Delay(self, timp):
        sleep(timp / 100000000.0) # ~ 10 ns

    def __LineSelect(self, line):
        if line >= LINES:
            return

        b = [0] * 4

        for i in range(0, 4):
            b[i] = line >> i
            b[i] = b[i] & 0x1

        GPIO.output(self.__D, b[3])
        GPIO.output(self.__C, b[2])
        GPIO.output(self.__B, b[1])
        GPIO.output(self.__A, b[0])


    def Draw(self, array):

        for line in range(0, LINES):
            GPIO.output(self.__G, 0)
            GPIO.output(self.__LAT, 0)
            GPIO.output(self.__CLK, 0)
            
            self.__LineSelect(line)

            for column in range(0, LINES):
                #self.Delay(1)

                GPIO.output(self.__DI, array[line][column])

                #self.Delay(1)
                GPIO.output(self.__CLK, 1)

                #self.Delay(1)
                GPIO.output(self.__CLK, 0)

            GPIO.output(self.__LAT, 1)
            #self.Delay(1)


    def __init__(self, D, C, B, A, G, DI, CLK, LAT):
        self.__D = D
        self.__C = C
        self.__B = B
        self.__A = A
        self.__G = G
        self.__DI = DI
        self.__CLK = CLK
        self.__LAT = LAT

        GPIO.setmode(GPIO.BOARD)

        GPIO.setup(D, GPIO.OUT)
        GPIO.setup(C, GPIO.OUT)
        GPIO.setup(B, GPIO.OUT)
        GPIO.setup(A, GPIO.OUT)
        GPIO.setup(G, GPIO.OUT)
        GPIO.setup(DI, GPIO.OUT)
        GPIO.setup(CLK, GPIO.OUT)
        GPIO.setup(LAT, GPIO.OUT)

def main():
    if len(sys.argv) is not 9:
        print("Faulty pin parameters - D C B A G DI CLK LAT")
        exit(1)

    led = Matrix(
        int(sys.argv[1]),
        int(sys.argv[2]),
        int(sys.argv[3]),
        int(sys.argv[4]),
        int(sys.argv[5]),
        int(sys.argv[6]),
        int(sys.argv[7]),
        int(sys.argv[8])
    )

    while True:
        led.Draw(matrix)
        led.Delay(1)

main()

File: cpp/matrix.cpp
#include<iostream>
#include<string>
#include<wiringPi.h>
#include<time.h>
#include "matrix.h"

using namespace std;

#define LINES 16
#define HIGH 1
#define LOW 0

// 1 - OFF, 0 - ON
int matrix[LINES][LINES] = {
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0},
    {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0}
};

Matrix::Matrix(int D, int C, int B, int A, int G, int DI, int CLK, int LAT)
{
    this->D = D;
    this->C = C;
    this->B = B;
    this->A = A;
    this->G = G;
    this->DI = DI;
    this->CLK = CLK;
    this->LAT = LAT;

    wiringPiSetupPhys();

    pinMode(D, OUTPUT);
    pinMode(C, OUTPUT);
    pinMode(B, OUTPUT);
    pinMode(A, OUTPUT);
    pinMode(G, OUTPUT);
    pinMode(DI, OUTPUT);
    pinMode(CLK, OUTPUT);
    pinMode(LAT, OUTPUT);
}

void Matrix::Delay(long int nanoseconds)
{
    delayMicroseconds(nanoseconds / 100); // ~ 10 ns
}

void Matrix::LineSelect(int line)
{
    if(line > LINES)
        return;
    
    int b[4];
    
    for (int i = 0; i < 4; i++)
    {
        b[i] = param >> i;
        b[i] = b[i] & 0x1;
    }

    digitalWrite(D, b[3]);
    digitalWrite(C, b[2]);
    digitalWrite(B, b[1]);
    digitalWrite(A, b[0]);
}

void Matrix::Draw(int draw[16][16])
{
    for (int line = 0; line < LINES; line++)
    {
        digitalWrite(G, LOW);
        digitalWrite(LAT, LOW);
        digitalWrite(CLK, LOW);

        for (int column = 0; column < LINES; column++)
        {
            //Delay(1);

            digitalWrite(DI, draw[line][column]);

            //Delay(1);
            digitalWrite(CLK, 1);

            //Delay(1);
            digitalWrite(CLK, 0);
        }

        digitalWrite(LAT, 1);
        //Delay(1);
    }
}

int i(char * param)
{
    return strtol(param, nullptr, 10);
}

int main(int argc, char** argv)
{
if (argc != 9)
    {
        cout << "Faulty pin parameters - D C B A G DI CLK LAT";
        return 1;
    }

    Matrix led = Matrix(
        i(argv[1]),
        i(argv[2]),
        i(argv[3]),
        i(argv[4]),
        i(argv[5]),
        i(argv[6]),
        i(argv[7]),
        i(argv[8])
    );

    while (true)
    {
        led.Draw(matrix);
        led.Delay(1);
    }
}